piecrust/src/store.rs

```diff
@@ -0,0 +1,903 @@
+    let root_dir = root_dir.as_ref();
+    let mut commits = BTreeMap::new();
+
+    for entry in fs::read_dir(root_dir)? {
+        let entry = entry?;
+        if entry.path().is_dir() {
+            let (root, commit) = read_commit(entry.path())?;
+            commits.insert(root, commit);
+        }
+    }
+
+    Ok(commits)
+}
+
+fn read_commit<P: AsRef<Path>>(commit_dir: P) -> io::Result<(Root, Commit)> {
+    let commit_dir = commit_dir.as_ref();
+
+    let root = root_from_dir(commit_dir)?;
+    let commit = commit_from_dir(commit_dir)?;
+
+    Ok((root, commit))
+}
+
+fn root_from_dir<P: AsRef<Path>>(dir: P) -> io::Result<Root> {
+    let dir = dir.as_ref();
+    let dir_name = dir.file_name().unwrap().to_str().ok_or_else(|| {
+        io::Error::new(
+            io::ErrorKind::InvalidData,
+            format!("Directory name \"{dir:?}\" is invalid UTF-8"),
+        )
+    })?;
+
+    let dir_name_bytes = hex::decode(dir_name).map_err(|err| {
+        io::Error::new(
+            io::ErrorKind::InvalidData,
+            format!("Directory name \"{dir_name}\" is invalid hex: {err}"),
+        )
+    })?;
+
+    if dir_name_bytes.len() != ROOT_LEN {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            format!("Expected directory name \"{dir_name}\" to be of size {ROOT_LEN}, was {}", dir_name_bytes.len())
+        ));
+    }
+
+    let mut root = [0u8; ROOT_LEN];
+    root.copy_from_slice(&dir_name_bytes);
+
+    Ok(root)
+}
+
+fn commit_from_dir<P: AsRef<Path>>(dir: P) -> io::Result<Commit> {
+    let dir = dir.as_ref();
+
+    let merkle_path = dir.join(MERKLE_FILE);
+
+    let modules = merkle_from_path(merkle_path)?;
+    let mut diffs = BTreeSet::new();
+
+    let bytecode_dir = dir.join(BYTECODE_DIR);
+    let memory_dir = dir.join(MEMORY_DIR);
+
+    for module in modules.keys() {
+        let module_hex = hex::encode(module);
+
+        // Check that all modules in the merkle file have a corresponding
+        // bytecode and memory.
+        let bytecode_path = bytecode_dir.join(&module_hex);
+        if !bytecode_path.is_file() {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                format!("Non-existing bytecode for module: {module_hex}"),
+            ));
+        }
+
+        let memory_path = memory_dir.join(&module_hex);
+        if !memory_path.is_file() {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                format!("Non-existing memory for module: {module_hex}"),
+            ));
+        }
+
+        // If there is a diff file for a given module, register it in the map.
+        let diff_path = memory_path.with_extension(DIFF_EXTENSION);
+        if diff_path.is_file() {
+            diffs.insert(*module);
+        }
+    }
+
+    Ok(Commit { modules, diffs })
+}
+
+fn merkle_from_path<P: AsRef<Path>>(
+    path: P,
+) -> io::Result<BTreeMap<ModuleId, Root>> {
```

Comment:
I'd consider changing the name of this function as it is misleading, it returns a map of module roots rather than a single Merkle root